lms/services/canvas_api/client.py

```diff
@@ -226,16 +227,31 @@ class _ListFilesSchema(RequestsResponseSchema):
         id = fields.Integer(required=True)
         updated_at = fields.String(required=True)
 
-    def public_url(self, file_id):
+    def public_url(self, course_id, file_id, is_instructor):
         """
         Get a new temporary public download URL for the file with the given ID.
 
+        :param course_id: the ID of the Canvas course
+        :rtype: str
+
         :param file_id: the ID of the Canvas file
         :rtype: str
+
+        :param is_instructor: Whether the current user is an instructor
+        :type is_instructor: bool
+
+        :raise CanvasFileNotFoundInCourse: If there's no file with ID file_id
+            in the course with ID course_id
         """
         # For documentation of this request see:
         # https://canvas.instructure.com/doc/api/files.html#method.files.public_url
 
+        if is_instructor:
+            if not any(
+                str(file_["id"]) == file_id for file_ in self.list_files(course_id)
+            ):
```

Comment:
The use of `any()` like this avoids iterating over all of the files returned by `self.list_files(course_id)` unless necessary. If `list_files()` were a generator this could avoid unnecessary Canvas API requests (but it's not).

We do `str(file_["id"])` on each file because anything can be turned into a string. If we instead did `int(file_id)` on the one `file_id` variable, we'd have to deal with `int()` potentially crashing if a non-int `file_id` were passed to the API

Reply:
We call the Canvas API to get a list of all the files in the course, with possible pagination and multiple requests, in order to determine whether the file is in the course.

We only want to do this for instructors because:

* We don't want to add these additional Canvas API calls to every Canvas Files assignment launch. Only to instructor launches
* The exception that we raise triggers and error page that tells the user to edit the assignment and re-select the file. Only instructors can do that
* We can't actually tell whether the file is in the course when a student launches an assignment anyway. The list files API will only list the files that the student has access to, which doesn't necessarily include all of the files in the course (instructors can control whether files are visible to students or not)

When a student launches an assignment if they don't have access to the file (which could be for a variety of reasons) we get a permissions error from the Canvas API and we show an error message to the student. That was done in a previous PR: https://github.com/hypothesis/lms/pull/2223

Reply:
Do you have an idea about how much overhead this will add to instructor launches? I guess we could cache the file ID => course ID association, although that doesn't account for a file being deleted. If the file _is_ deleted however, then the subsequent call to get a URL for the file itself will presumably fail. Alternatively we could remove this check from this API and instead make the frontend make a separate request which can happen in parallel with fetching the file URL. If that second call eventually fails the frontend can switch to the error state.

Reply:
I think we can cache them in the DB, but that needs to be verified. As far as I know the course that a given file ID is associated with cannot change in Canvas. I don't see any way to move files between courses, for example. And course copy creates new copies of the files with new IDs. But I think we'd need to verify with greater certainty that a file ID always belongs to the same course. If so we could cache this result in the DB

Reply:
Listing a site's files seems to take anything from 0.5 sec to 3 secs. It looks like the ceiling is more like about 1.5 secs though, with the slower ones being outliers: https://one.nr/0OwvO83R

Reply:
We certainly could move this check to a separate API endpoint and that might speed things up as the list files request to Canvas could be happening in parallel with the public URL request. I'm not sure whether I think it's worthwhile or not. It might actually make the code a bit nicer. Something to try in a follow-up PR maybe.

Since the frontend already has the `file_id`, this could be a frontend-only change: it could call the existing list files API and check for the ID itself. The frontend would need to know whether the user was an instructor, which I'm not sure it does currently. But I think that would close off the possibility of caching the results in the DB in future.